pageserver/src/layered_repository/timeline.rs

```diff
@@ -749,7 +749,7 @@ impl Timeline {
         }
 
         if task_guard.is_none() {
-            let thread_timeline = Arc::clone(self);
+            let thread_timeline = Arc::downgrade(self);
```

Comment:
Why this change?

I don't think this Weak reference is really worth it. Yeah, if the timeline is deleted right before the timeline calculation starts in the other thread, it's pointless to continue with the calculation and it would indeed be good to stop it as early as possible. But the window is really narrow between spawning the new thread, and the new thread upgrading the reference again. It would be more meaningful to sprinkle `is_shutdown_requested()` checks in `calculate_logical_size`, so that it would exit more quickly if shutdown is requested. But that seems unrelated to the rest of the PR.

Reply:
I've managed to reproduce it once during my frantic test reruns and without this change it looks very alarming: we do log that some timeline was deleted and later we do log that this timeline has its size calculated.

I can sure revert this, but feel quite uneasy about that. I'm not sure that `is_shutdown_requested()` is good generally, since it's up to the developer to check that.
We're clearly able to see where such approach lead us right here 🙂 
`Weak` as an alternative is something that we're forced to check which makes it more safe, IMO.

Reply:
Hmm, 'delete_timeline` calls `shutdown_threads` before performing the deletion. That's supposed to wait for all background threads to finish. Did it not wait for the initial timeline size calculation to finish? (/me looks at code). `try_spawn_size_init_task` doesn't use the `thread_mgr.rs` facility, so indeed that helper thread is not registered with the thread manager, and `shutdown_threads` would not wait for it. That seems wrong, but it's not new with this PR.

Reply:
We could replace the thread registry mechanism and `shutdown_threads` with something based on Rust lifetimes instead. Not just here, but in general. Like, instead of waiting for the threads to exit, wait for all strong Arc references to the Timeline to be dropped. It's a bit more complicated if you want to wait for just some threads to finish, e.g. if you want to wait for the WAL receiver to finish. Then the WAL receiver would need to hold some kind of a "baton" object or wrapper around Timeline, so that you could wait for that to be dropped instead of the whole Timeline.

(and with https://github.com/neondatabase/neon/pull/2376, all of the above still applies, but replace thread with task).

Reply:
Yes, there's definitely something in that idea that I like now, but that would require more changes everywhere, since now we just toss `Arc`s without thinking much.
We would need to create some newtypes and straighten the API better so nobody is able to accidentally "steal" the entire `Arc` into some isolated task that never drops.

Definitely a separate PR, will extract this small change too.

Reply:
I've reverted the changes and not bringing them up in a separate PR for now: looks like clones of `Arc<Timeline>` slips away in some contexts like `gc_loop` and I'm not sure how to join on a thread after registration, due to the comment on `thread_mgr::register` function:

> /// NOTE: threads registered through this function cannot be joined

So, I've decided to leave with it for now, before big timeline overhaul that's supposed to happen eventually.